fix: navigation to correct heading in tools section

Vishal2002 opened the pull request:
Description

- This PR resolves the issue of failing to navigate to the correct link in the tool section.
- Utilized the `useRef` hook to reference the target element, improving navigation reliability.
Related issue(s)
Fixes: #2181 

https://github.com/asyncapi/website/assets/35897449/0b00c9fc-49c4-4755-9a37-1e735bfc19a2

Vishal2002 committed:
Navigation to correct heading in tools section

sambhavgupta0705 commented:
@Vishal2002 please the correct pr title convention

Vishal2002 commented:
> @Vishal2002 please the correct pr title convention

@sambhavgupta0705  Done Sir

Vishal2002 committed:
Merge branch 'master' into scroll-issue

sambhavgupta0705 commented:
@Vishal2002 it doesn't solve the issue for us
As we need the url to be responsive to the request

Vishal2002 commented:
> @Vishal2002 it doesn't solve the issue for us As we need the url to be responsive to the request

But as per an actual issue by @derberg where he mentioned that when trying to navigate to https://www.asyncapi.com/tools#Documentation%20Generators he is unable to navigate to the actual heading.After that I also read the discussion where @akshatnema told a fellow contributor to use `useref `and `scroll into View` to solve the issue.I did the same and solved it. Would be great if you give some more context about my mistake.

sambhavgupta0705 commented:
ohhh my bad
I missed the preview you added 😅

sambhavgupta0705 commented on components/tools/ToolsList.js:
```diff
@@ -1,12 +1,25 @@
+
+  useEffect(() => {
+    const hash = window.location.hash;
+    if (hash) {
```
can  you  please explain  what does `hash` do internally

Vishal2002 replied:
So basically in `hash` we are storing the fragment of url following`#` .

![Screenshot 2024-03-26 123216](https://github.com/asyncapi/website/assets/35897449/ad1ab558-418d-4ee7-b07e-374667540e3c)

sambhavgupta0705 commented on components/tools/ToolsList.js:
```diff
@@ -1,12 +1,25 @@
+    const hash = window.location.hash;
+    if (hash) {
+      const decodedHash = decodeURIComponent(hash.slice(1));
+      const categoryRef = categoryRefs.current[decodedHash];
```
`decodedHash` checks the current URL entered??

Vishal2002 replied:
`decodedHash` have purified query in url , like sometimes the spaces in the URL becomes encoded into special characters and ASCII characters like % and %20. So to decode them we use decodeURIComponent.

sambhavgupta0705 reviewed (commented):
@Vishal2002 I have checked the netlify preview it works fine 
Just added some comments for better understanding

sambhavgupta0705 commented:
@Vishal2002 any update??

Vishal2002 commented:
> @Vishal2002 any update??

Sir already answered the question which you asked about `hash` and `decodeHash`.

sambhavgupta0705 commented:
I can't see any reply
I think you haven't finished the review

Vishal2002 commented:
> I can't see any reply I think you haven't finished the review

Sorry, ya I failed to resolve that, I think now it will be visible to you.

sambhavgupta0705 commented:
still cannot see 
you marked them as resolved only

Vishal2002 commented:
> still cannot see you marked them as resolved only

IDK why this is happening but I have commented on the question that you asked me before about` hash` and `decodehash`.

![Screenshot 2024-03-30 160825](https://github.com/asyncapi/website/assets/35897449/0e7899de-0d71-4413-94d1-5e0f8536d337)
![Screenshot 2024-03-30 160813](https://github.com/asyncapi/website/assets/35897449/3f00cc75-35f2-48db-a853-4f535001d19f)

akshatnema commented:
> IDK why this is happening but I have commented on the question that you asked me before about hash and decodehash.

@Vishal2002 I can see that your review is still pending. Inside the `Files changed`, click on the `Review Changes` button and click `comment` option there. Your review will be posted.

Vishal2002 commented:
> > IDK why this is happening but I have commented on the question that you asked me before about hash and decodehash.
> 
> @Vishal2002 I can see that your review is still pending. Inside the `Files changed`, click on the `Review Changes` button and click `comment` option there. Your review will be posted.

Thank you @akshatnema  Sir and @sambhavgupta0705  Sorry for the inconvenience.

akshatnema commented on components/tools/ToolsList.js:
```diff
@@ -1,12 +1,25 @@
+        categoryRef.scrollIntoView({ behavior: 'smooth', block: 'start' });
+      }
+    }
+  }, []);
```
```suggestion
  }, []);
  
```

Vishal2002 replied:
What changes to add here ?

sambhavgupta0705 replied:
just add a newline at the end of the file

sambhavgupta0705 replied:
clean code rules

Vishal2002 replied:
Done Sir

akshatnema commented on components/tools/ToolsList.js:
```diff
@@ -1,12 +1,25 @@
 export default function toolsList({ toolsData }) {
+  const categoryRefs = useRef({});
+
+  useEffect(() => {
```
Add a comment, explaining about this `useEffect` block.

Vishal2002 replied:
It scrolls the page to a specific category referenced in the URL hash.

Vishal2002 replied:
@akshatnema  Done Sir

Vishal2002 committed:
clean code and comment regarding useEffect

Vishal2002 committed:
Merge branch 'scroll-issue' of https://github.com/Vishal2002/website into scroll-issue

akshatnema committed:
Merge branch 'master' into scroll-issue

anshgoyalevil committed:
Merge branch 'master' into scroll-issue

sambhavgupta0705 reviewed (approved):
LGTM

anshgoyalevil committed:
Merge branch 'master' into scroll-issue

anshgoyalevil commented:
@Vishal2002
Great work!
The solution seems to be working but there's a little bit of extra scroll to the view when the page is loaded.

Example:
[screen-capture (7).webm](https://github.com/asyncapi/website/assets/94157520/3e55989d-fc79-4865-8ede-f4b16399eacd)

sambhavgupta0705 commented:
@anshgoyalevil it is working fine for me 
Can you please recheck it on your end  and then we are good tyo merge this PR

akshatnema commented:
@sambhavgupta0705 @anshgoyalevil I find that there is extra scroll on mobile view. Can you check once?

anshgoyalevil commented:
> Can you please recheck it on your end and then we are good tyo merge this PR

There's still extra scroll on the desktop and mobile view. Have room for improvement.
Try opening the link https://deploy-preview-2790--asyncapi-website.netlify.app/tools#Others 

> @sambhavgupta0705 @anshgoyalevil I find that there is extra scroll on mobile view. Can you check once?

Yes, it is there for me as well.

Vishal2002 commented:
> > Can you please recheck it on your end and then we are good tyo merge this PR
> 
> There's still extra scroll on the desktop and mobile view. Have room for improvement. Try opening the link https://deploy-preview-2790--asyncapi-website.netlify.app/tools#Others
> 
> > @sambhavgupta0705 @anshgoyalevil I find that there is extra scroll on mobile view. Can you check once?
> 
> Yes, it is there for me as well.

I didn't find this mobile view issue, Sir please can you tell me how to produce this again?
https://drive.google.com/file/d/1haMMTh7psG3pICxCwDqKHd4IfwLBzZBD/view?usp=sharing

anshgoyalevil commented:
> > > Can you please recheck it on your end and then we are good tyo merge this PR
> > 
> > 
> > There's still extra scroll on the desktop and mobile view. Have room for improvement. Try opening the link https://deploy-preview-2790--asyncapi-website.netlify.app/tools#Others
> > > @sambhavgupta0705 @anshgoyalevil I find that there is extra scroll on mobile view. Can you check once?
> > 
> > 
> > Yes, it is there for me as well.
> 
> I didn't find this mobile view issue, Sir please can you tell me how to produce this again? https://drive.google.com/file/d/1haMMTh7psG3pICxCwDqKHd4IfwLBzZBD/view?usp=sharing

In this video itself, you can see the extra scroll issue. From 00:00 to 00:15 seconds.

Vishal2002 commented:
> > > > Can you please recheck it on your end and then we are good tyo merge this PR
> > > 
> > > 
> > > There's still extra scroll on the desktop and mobile view. Have room for improvement. Try opening the link https://deploy-preview-2790--asyncapi-website.netlify.app/tools#Others
> > > > @sambhavgupta0705 @anshgoyalevil I find that there is extra scroll on mobile view. Can you check once?
> > > 
> > > 
> > > Yes, it is there for me as well.
> > 
> > 
> > I didn't find this mobile view issue, Sir please can you tell me how to produce this again? https://drive.google.com/file/d/1haMMTh7psG3pICxCwDqKHd4IfwLBzZBD/view?usp=sharing
> 
> In this video itself, you can see the extra scroll issue. From 00:00 to 00:15 seconds.

The additional scrolling occurs infrequently.

sambhavgupta0705 commented:
> > > > > Can you please recheck it on your end and then we are good tyo merge this PR
> > > > 
> > > > 
> > > > There's still extra scroll on the desktop and mobile view. Have room for improvement. Try opening the link https://deploy-preview-2790--asyncapi-website.netlify.app/tools#Others
> > > > > @sambhavgupta0705 @anshgoyalevil I find that there is extra scroll on mobile view. Can you check once?
> > > > 
> > > > 
> > > > Yes, it is there for me as well.
> > > 
> > > 
> > > I didn't find this mobile view issue, Sir please can you tell me how to produce this again? https://drive.google.com/file/d/1haMMTh7psG3pICxCwDqKHd4IfwLBzZBD/view?usp=sharing
> > 
> > 
> > In this video itself, you can see the extra scroll issue. From 00:00 to 00:15 seconds.
> 
> The additional scrolling occurs infrequently.

The extra scroll is happening for me also
@Vishal2002 we need to fix it before merging this one

Vishal2002 commented:
> > > > > > Can you please recheck it on your end and then we are good tyo merge this PR
> > > > > 
> > > > > 
> > > > > There's still extra scroll on the desktop and mobile view. Have room for improvement. Try opening the link https://deploy-preview-2790--asyncapi-website.netlify.app/tools#Others
> > > > > > @sambhavgupta0705 @anshgoyalevil I find that there is extra scroll on mobile view. Can you check once?
> > > > > 
> > > > > 
> > > > > Yes, it is there for me as well.
> > > > 
> > > > 
> > > > I didn't find this mobile view issue, Sir please can you tell me how to produce this again? https://drive.google.com/file/d/1haMMTh7psG3pICxCwDqKHd4IfwLBzZBD/view?usp=sharing
> > > 
> > > 
> > > In this video itself, you can see the extra scroll issue. From 00:00 to 00:15 seconds.
> > 
> > 
> > The additional scrolling occurs infrequently.
> 
> The extra scroll is happening for me also @Vishal2002 we need to fix it before merging this one

@sambhavgupta0705  Sir at my side it happens for the first time when the page renders and after that, everything go smoothly.

anshgoyalevil commented:
@Vishal2002 

> happens for the first time when the page renders

That's what we are targeting in this issue. An end user would click on a link that would render for the first time for them. They won't be stuffing the hash values to URLs manually after the first render

Vishal2002 commented:
> @Vishal2002
> 
> > happens for the first time when the page renders
> 
> That's what we are targeting in this issue. An end user would click on a link that would render for the first time for them. They won't be stuffing the hash values to URLs manually after the first render

@anshgoyalevil and @sambhavgupta0705  Sir there might be issue with height of the container and if we add height:100vh  then it solves the issue.


https://github.com/asyncapi/website/assets/35897449/e005b742-55d1-4953-ab45-2e4c5c88c00e

Vishal2002 commented:
@anshgoyalevil and @sambhavgupta0705 Sir looking for the reply from your side.

sambhavgupta0705 commented:
@Vishal2002 push that change once to this branch and let us see the preview

Vishal2002 committed:
fix:added container height to resolve extra scroll

added container height to resolve extra scroll

added container height to resolve extra scroll

added height to resolve the extra scroll issue

Vishal2002 committed:
fix:added container height to resolve extra scroll

Vishal2002 committed:
added container height to resolve extra scroll

Vishal2002 committed:
added container height to resolve extra scroll

Vishal2002 committed:
added height to resolve the extra scroll issue

Vishal2002 committed:
Merge branch 'scroll-issue' of https://github.com/Vishal2002/website into scroll-issue

Vishal2002 committed:
addding height to container

sambhavgupta0705 commented:
@Vishal2002 the issue is still there

sambhavgupta0705 commented:
@Vishal2002 please resolve the merge conflicts

Vishal2002 commented:
@sambhavgupta0705 Ok and  the previous issue regarding extra scroll might be because of useref unable to get the reference for the first time the component mounts.
